linode/resource_linode_user.go

```diff
@@ -107,3 +243,140 @@ func resourceLinodeUserDelete(ctx context.Context, d *schema.ResourceData, meta
 	}
 	return nil
 }
+
+func updateUserGrants(ctx context.Context, d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ProviderMeta).Client
+
+	username := d.Get("username").(string)
+	restricted := d.Get("restricted").(bool)
+
+	if !isUserGrantSchemaUpdated(d) {
+		return nil
+	}
+
+	if !restricted {
+		return fmt.Errorf("user must be restricted in order to update grants")
+	}
+
+	updateOpts := linodego.UserGrantsUpdateOptions{}
+
+	if global, ok := d.GetOk("global_grants"); ok {
+		global := global.([]interface{})[0].(map[string]interface{})
+		updateOpts.Global = expandGrantsGlobal(global)
+	}
+
+	updateOpts.Domain = expandGrantsEntities(d.Get("domain_grant").(*schema.Set).List())
+	updateOpts.Image = expandGrantsEntities(d.Get("image_grant").(*schema.Set).List())
+	updateOpts.Linode = expandGrantsEntities(d.Get("linode_grant").(*schema.Set).List())
+	updateOpts.Longview = expandGrantsEntities(d.Get("longview_grant").(*schema.Set).List())
+	updateOpts.NodeBalancer = expandGrantsEntities(d.Get("nodebalancer_grant").(*schema.Set).List())
+	updateOpts.StackScript = expandGrantsEntities(d.Get("stackscript_grant").(*schema.Set).List())
+	updateOpts.Volume = expandGrantsEntities(d.Get("volume_grant").(*schema.Set).List())
+
+	if _, err := client.UpdateUserGrants(ctx, username, updateOpts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// isUserGrantSchemaUpdated returns whether or not any grant-related field is updated.
+func isUserGrantSchemaUpdated(d *schema.ResourceData) bool {
+	result := false
+	for key := range resourceLinodeUser().Schema {
```

Comment:
We should probably keep all user grant-related field names in a static string slice for use in multiple places. That would read much cleaner than dynamically filtering fields in the schema everytime